### DISCUSSION TITLE
Cannot calibrate NV12 without a LUT?

### DISCUSSION THREAD

**Discussion by chcore** [awawa-dev/HyperHDR#997] [Q&A]
I am testing out the NV12 support in master, and noticed that HyperHDR will not accept NV12 input unless there is a LUT in the home directory.

I can of course put a random LUT in there so that it at least works, but then wouldn't the calibration process be tainted since the source is using a bad or incorrect LUT?

Side note, the "test_HDR_yuv420_low_quality.mp4" does not trigger HDR mode on my TV when played through Kodi. Is it not supposed to trigger HDR mode?

**Comment by awawa-dev** [comment]
Hi

Please use https://github.com/awawa-dev/HyperHDR/actions/runs/12069667142 installers from https://github.com/awawa-dev/HyperHDR/pull/996

> Side note, the "test_HDR_yuv420_low_quality.mp4" does not trigger HDR mode on my TV when played through Kodi. Is it not supposed to trigger HDR mode?

Yes, it should. Unless your TV is already in the HDR mode. On LG you can verify it in the currect video source info.

**Reply by awawa-dev** [reply]
> > Side note, the "test_HDR_yuv420_low_quality.mp4" does not trigger HDR mode on my TV when played through Kodi. Is it not supposed to trigger HDR mode?
> 
> Yes, it should. Unless your TV is already in the HDR mode. On LG you can verify it in the currect video source info.

Not every player can handle full range yuv 420. Try limited range yuv 420 test file.

**Reply by chcore** [reply]
> > > Side note, the "test_HDR_yuv420_low_quality.mp4" does not trigger HDR mode on my TV when played through Kodi. Is it not supposed to trigger HDR mode?
> > 
> > 
> > Yes, it should. Unless your TV is already in the HDR mode. On LG you can verify it in the currect video source info.
> 
> Not every player can handle full range yuv 420. Try limited range yuv 420 test file.

That file doesn't trigger HDR mode for me in Kodi either. This is on linux/GBM/intel/x86.

I generally have no issues with triggering HDR mode with other content.

**Reply by chcore** [reply]
None of the HDR calibration videos switch the TV to HDR mode in Kodi. What's the format of these? I think Kodi on linux only supports static formats (HLG, HDR10).

**Reply by awawa-dev** [reply]
You have a command that created these test HDR10 & SDR files in the same folder in the source. If limited yuv420 hdr doesnt trigger HDR then definitely something wrong is with the player because even built-in LG player can plan this file.

**Reply by chcore** [reply]
> You have a command that created these test HDR10 & SDR files in the same folder in the source

Good tip. I am able to generate a file that triggers the TV's HDR mode in Kodi by adding the x265 param: 

`master-display=G(8500,39850)B(6550,23000)R(35400,15650)WP(15635,16450)L(10000000,1)WP(15635,16450)L(1000000,100%)`

But I won't pretend to know what that does, or whether it will introduce inaccuracies into the calibration process?

**Reply by awawa-dev** [reply]
I don't know either ;) Theoretically the hdr10 tag should be enough and it works in the LG player, Windows players with OS HDR enabled (madvr codec, vlc: these will play everything even yuv444) and from what I know also on nvidia shield (but only HDR limited yuv420). If this new parametert modifies the target brightness then you need to check in the logs how much it came out. It should be 200 nits in the calibration process.

**Reply by chcore** [reply]
Is the calibration process deterministic? That is, if I run it twice on the same test video, are the LUTs going to be identical? I saw in a pull request there was some AI component?

**Reply by awawa-dev** [reply]
The process is completely deterministic, but the source may cause interference, e.g. random dither when decoding yuv420 if you are not able to disable it (or the webos capturing/grabber noise): this may affect the result marginally. As I noted in the comments, due to the number and nature of the changes that now allow calibrating the player without using a browser, AI has been postponed to the next stage.

**Reply by chcore** [reply]
Possibly related: https://github.com/xbmc/xbmc/pull/24160

**Reply by awawa-dev** [reply]
Yes, it's about what's in that PR in Kodi and it looks like they don't intend to merge it despite the fact that other players don't block HDR passthrough for such and similar files. The problem with adding this meta-tag you're writing about is that the current test files have already been tested by many people on different environments/players and we would have to start everything from scratch.

This is a very malicious Kodi problem because if you didn't notice that the TV wasnt in HDR mode, you would have HyperHDR calibrated for ... SDR. I don't know, maybe I will add a subdirectory called Kodi with test files with this meta-tag?

**Reply by chcore** [reply]
I compiled a build with that patch, and can confirm that the test video now triggers HDR mode as expected.

> I don't know, maybe I will add a subdirectory called Kodi with test files with this meta-tag?

Probably worth doing if there are as many Kodi users using HyperHDR as I think there are.

**Reply by awawa-dev** [reply]
> master-display=G(8500,39850)B(6550,23000)R(35400,15650)WP(15635,16450)L(10000000,1)WP(15635,16450)L(1000000,100%)

But I don't recognize these color primaries. Is this a characteristic of some TV/projector?
I'm afraid that if I supply some constant here... on the other hand, if it's just passthrough data, then everything can be messed up is the image displayed on the TV. Unless the grabber for some reason reads these metatags and applies them as well.

**Reply by awawa-dev** [reply]
> master-display=G(8500,39850)B(6550,23000)R(35400,15650)WP(15635,16450)L(10000000,1)WP(15635,16450)L(1000000,100%)

oh, it's such a strange format and after deciphering:
R=(0.708 , **0.313**)
B=(0.131 , **0.46**)
G=(0.17 , 0.797)

if not such a bizarre setting for blue it could be a Rec.2020

**Reply by awawa-dev** [reply]
For rec.2020 it should be:
`master-display=G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,1)`
@chcore  could you verify if that setting also triggers HDR for Kodi?

**Reply by chcore** [reply]
> @chcore could you verify if that setting also triggers HDR for Kodi?

Yes it does

**Reply by chcore** [reply]
> But I don't recognize these color primaries. Is this a characteristic of some TV/projector?

I copied it from some repository/gist on Github while searching for HDR-related parameters I could try with ffmpeg. I suspect that setting these to any values at all will trigger HDR mode (albeit with incorrect colors).

**Reply by awawa-dev** [reply]
I've just tested it on RPI4 using the latest available on bullseye Kodi 19.1 and an old MP4 test file (without the `master-display` tag) and it works: HDR mode is triggered on LG TV. Maybe they've fixed this already?

Edit: Bookworm and Kodi 20.5 also works fine.

**Reply by chcore** [reply]
According to one of the comments on  that pull request, the bug does not affect RPI, only Intel:

> Not to mention that DRM PRIME platforms, eg. RPI4 allow switching to HDR mode with those "incomplete metadata" videos. Why this should be different for Intel?

**Reply by awawa-dev** [reply]
OK, I'll test it using my N100 mini PC.

**Reply by awawa-dev** [reply]
Do you think librelec x64 should be enough?

**Reply by chcore** [reply]
> Do you think librelec x64 should be enough?

Probably, but I wouldn't be surprised if they have already merged this patch or something like it

**Reply by awawa-dev** [reply]
OK, I installed Libreelec on N100 and after a lot of trouble (I couldn't enable HDR for any files => I had to manually switch the resolution to 1080p in Kodi) I managed to get the Intel bug you found.
I updated both HDR files with a new x265 master-display tag: https://github.com/awawa-dev/awawa-dev.github.io/tree/master/calibration. I'll probably focus on 420 limited in the future, but the old files are still intact in the 'old' subfolder.

**Reply by chcore** [reply]
Thanks, the new HDR test file triggers HDR mode as expected.

**Reply by chcore** [reply]
Do the "color calibration" settings in the UI affect the output of the LUT generation/calibration process?

**Reply by awawa-dev** [reply]
Yes: https://github.com/awawa-dev/HyperHDR/wiki/LUT-calibration You can compare score (error) and calibrated main colors values with and without that option enabled. Only P010 result seems to be unaffected but its already very good without it.

**Reply by chcore** [reply]
> Yes: https://github.com/awawa-dev/HyperHDR/wiki/LUT-calibration You can compare score (error) and calibrated main colors values with and without that option enabled. Only P010 result seems to be unaffected but its already very good without it.

Sorry, I meant these settings:

![image](https://github.com/user-attachments/assets/557c38d0-2594-4599-9698-c16fdfee02c5)

**Reply by awawa-dev** [reply]
No, these LED settings does not affect the LUT calibration.

**Comment by chcore** [comment]
Thanks @awawa-dev I see that build generates a "temporary" LUT when it cannot find one in the home directory. A couple of quick questions:

1. Is it beneficial to have separate LUTs for SDR and HDR?
2. If so, can I use the same temporary LUT while calibrating both SDR and HDR? Or must the temporary LUT be generated while the TV is in the same mode as the mode I'm trying to calibrate?

**Reply by awawa-dev** [reply]
1 for LG and piccap you need to have separate LUts for SDR and HDR, because webos processes these signals completely differently. And even what we get from SDR is not SDR anymore when captured so it needs LUT with its own transformation. This is result of the reverse engineering described in the PR development process here: https://github.com/awawa-dev/HyperHDR/pull/896 (quite a long thread, which needs to be expanded to see in its entirety)
2 one LUT will not work for both SDR and HDR at the same time on webos. SDR must have its own LUT (switched by webos HyperHDR backend) with tone mode enabled to improve the signal.
